torchrl/objectives/value/functional.py

```diff
@@ -948,46 +1098,47 @@ def _is_scalar(tensor):
                 "consider using the non-vectorized version of the return computation or splitting "
                 "your trajectories."
             )
-        else:
-            gammas[..., 1:, :] = gammas[..., 1:, :] * not_done.view(-1, 1, T, 1)
-
-    gammas_cp = torch.cumprod(gammas, -2)
-
-    lambdas = torch.ones(T + 1, 1, device=device)
-    lambdas[1:] = lmbda
-    lambdas_cp = torch.cumprod(lambdas, -2)
-
-    gammas = gammas[..., 1:, :]
-    lambdas = lambdas[1:]
 
-    dec = gammas_cp * lambdas_cp
-    if rolling_gamma in (None, True):
+    if rolling_gamma:
+        # Make the coefficient table
+        gammas = _make_gammas_tensor(gamma * not_done, T, rolling_gamma)
+        gammas_cp = torch.cumprod(gammas, -2)
+        lambdas = torch.ones(T + 1, 1, device=device)
+        lambdas[1:] = lmbda
+        lambdas_cp = torch.cumprod(lambdas, -2)
+        lambdas = lambdas[1:]
+        dec = gammas_cp * lambdas_cp
+
+        gammas = _make_gammas_tensor(gamma, T, rolling_gamma)
+        gammas = gammas[..., 1:, :]
         if gammas.ndimension() == 4 and gammas.shape[1] > 1:
             gammas = gammas[:, :1]
         if lambdas.ndimension() == 4 and lambdas.shape[1] > 1:
             lambdas = lambdas[:, :1]
-        v3 = (gammas * lambdas).squeeze(-1) * next_state_value
+
+        not_done = not_done.transpose(-2, -1).unsqueeze(-2)
+        if len(batch):
+            not_done = not_done.flatten(0, len(batch))
+        # lambdas = lambdas * not_done
+
+        v3 = (gammas * lambdas).squeeze(-1) * next_state_value * not_done
         v3[..., :-1] = 0
         out = _custom_conv1d(
-            reward + (gammas * (1 - lambdas)).squeeze(-1) * next_state_value + v3, dec
+            reward
+            + gammas.squeeze(-1)
+            * next_state_value
+            * (1 - lambdas.squeeze(-1) * not_done)
+            + v3,
+            dec,
```

Comment:
for bits like this I am not super able to check, so I am. trusting the tests

Reply:
Yeah this is not super clear, but essentially as long as the tests show that the result is equal to the non vectorized version we should be fine IMO